exec/call_test.go

```diff
@@ -0,0 +1,157 @@
+			Body: &fb,
+		},
+		{
+			Sig:        &fsig,
+			Native:     true,
+			NativeBody: reflect.ValueOf(testNativeFunction),
+		},
+	}
+
+	testNativeModule.Code = &wasm.SectionCode{
+		Bodies: []wasm.FunctionBody{fb},
+	}
+
+	// Once called, NewVM will execute the module's main
+	// function.
+	vm, err := NewVM(testNativeModule)
+	if err != nil {
+		t.Fatalf("Error creating VM: %v", vm)
+	}
+
+	if len(vm.funcs) < 1 {
+		t.Fatalf("Need at least a start function!")
+	}
+
+	// Only one entry, which should be a function
+	if secretVariable != secretValue {
+		t.Fatalf("x is %d instead of %d", secretVariable, secretValue)
+	}
+}
+
+var ModuleCallNative = []byte{
+	0x00, 0x61, 0x73, 0x6D, 0x01, 0x00, 0x00, 0x00, 0x01, 0x1A, 0x06, 0x60, 0x01, 0x7F, 0x00, 0x60,
+	0x01, 0x7F, 0x01, 0x7F, 0x60, 0x00, 0x01, 0x7F, 0x60, 0x00, 0x00, 0x60, 0x00, 0x01, 0x7C, 0x60,
+	0x01, 0x7F, 0x01, 0x7F, 0x02, 0x0F, 0x01, 0x03, 0x65, 0x6E, 0x76, 0x07, 0x5F, 0x6E, 0x61, 0x74,
+	0x69, 0x76, 0x65, 0x00, 0x05, 0x03, 0x02, 0x01, 0x02, 0x04, 0x04, 0x01, 0x70, 0x00, 0x02, 0x06,
+	0x10, 0x03, 0x7F, 0x01, 0x41, 0x00, 0x0B, 0x7F, 0x01, 0x41, 0x00, 0x0B, 0x7F, 0x00, 0x41, 0x01,
+	0x0B, 0x07, 0x09, 0x01, 0x05, 0x5F, 0x6D, 0x61, 0x69, 0x6E, 0x00, 0x01, 0x09, 0x01, 0x00, 0x0A,
+	0x08, 0x01, 0x06, 0x00, 0x41, 0x00, 0x10, 0x00, 0x0B,
+}
+
+func testNativeFunctionWithParameterAndReturnValue(x int32) int32 {
+	return x + 3
+}
+
+func importer(name string) (*wasm.Module, error) {
+
+	// A function signature. Both the native and WASM function
+	// have the same signature.
+	fsig := wasm.FunctionSig{
+		Form:        0, // XXX check that
+		ParamTypes:  []wasm.ValueType{wasm.ValueTypeI32},
+		ReturnTypes: []wasm.ValueType{wasm.ValueTypeI32},
+	}
+
+	// List of all function types available in this module.
+	// There is only one: (func [int32] -> [int32])
+	types := &wasm.SectionTypes{
+		Entries: []wasm.FunctionSig{fsig},
+	}
+
+	exports := &wasm.SectionExports{Entries: make(map[string]wasm.ExportEntry)}
+	exports.Entries["_native"] = wasm.ExportEntry{
+		FieldStr: "_native",
+		Kind:     wasm.ExternalFunction,
+		Index:    0,
+	}
+
+	testNativeModule := wasm.NewModule()
+	testNativeModule.Types = types
+	testNativeModule.FunctionIndexSpace = []wasm.Function{
+		{
+			Sig:        &fsig,
+			Native:     true,
+			NativeBody: reflect.ValueOf(testNativeFunctionWithParameterAndReturnValue),
+			Body:       &wasm.FunctionBody{},
+		},
+	}
+	testNativeModule.Export = exports
+
+	return testNativeModule, nil
+}
+
+func TestNativeSymbolCall(t *testing.T) {
+	m, err := wasm.ReadModule(bytes.NewReader(ModuleCallNative), importer)
+	if err != nil {
+		t.Fatalf("Could not read module: %v", err)
+	}
+	vm, err := NewVM(m)
+	if err != nil {
+		t.Fatalf("Could not instantiate vm: %v", err)
+	}
+	rtrns, err := vm.ExecCode(1)
+	if err != nil {
+		t.Fatalf("Error executing the default function: %v", err)
+	}
+	if int(rtrns.(uint32)) != 3 {
+		t.Fatalf("Did not get the right value. Expected %d, got %d", 3, rtrns)
```

Comment:
the convention (in Go, that I tried/try to follow in my projects) is to display first what we got, and then what we wanted.